staging/src/k8s.io/apiserver/pkg/endpoints/installer.go

```diff
@@ -68,6 +69,68 @@ type action struct {
 	AllNamespaces bool // true iff the action is namespaced but works on aggregate result for all namespaces
 }
 
+func ConvertGroupVersionIntoToDiscovery(list []metav1.APIResource) []apidiscoveryv2beta1.APIResourceDiscovery {
+	var apiResourceList []apidiscoveryv2beta1.APIResourceDiscovery
+	parentResources := map[string]*apidiscoveryv2beta1.APIResourceDiscovery{}
+
+	// Loop through all top-level resources
+	for _, r := range list {
+		if strings.Contains(r.Name, "/") {
+			// Skip subresources for now so we can get the list of resources
+			continue
+		}
+
+		var scope apidiscoveryv2beta1.ResourceScope
+		if r.Namespaced {
+			scope = apidiscoveryv2beta1.ScopeNamespace
+		} else {
+			scope = apidiscoveryv2beta1.ScopeCluster
+		}
+
+		apiResourceList = append(apiResourceList, apidiscoveryv2beta1.APIResourceDiscovery{
+			Resource: r.Name,
+			Scope:    scope,
+			ResponseKind: &metav1.GroupVersionKind{
+				Group:   r.Group,
+				Version: r.Version,
+				Kind:    r.Kind,
+			},
+			Verbs:      r.Verbs,
+			ShortNames: r.ShortNames,
+			Categories: r.Categories,
+		})
+		parentResources[r.Name] = &apiResourceList[len(apiResourceList)-1]
+	}
+
+	// Loop through all subresources
+	for _, r := range list {
+		// Split resource name and subresource name
+		split := strings.SplitN(r.Name, "/", 2)
+
+		if len(split) != 2 {
+			// Skip parent resources
+			continue
+		}
+
+		parent, exists := parentResources[split[0]]
+		if !exists {
+			// Had a subresource but not its parent???
```

Comment:
That's allowed?  Maybe I'm misreading, but that's the scenario Jordan and I were describing I think.  You need to create a synthetic parent with only the data we have, then fill it in incrementally.  You also need to handle errors when the scope doesn't agree.

Reply:
I understand the case that was described, but I don't that is not allowed in our installer code: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/installer.go#L220-L228 and also based on your comment: https://github.com/kubernetes/kubernetes/pull/111978#discussion_r972457385. Scope will also always agree because the same parent storage is accessed. golang apiservers that use this code should be bound by the condition.

If custom aggregated apiservers would like to create their own subresources without a parent, i think it makes more sense to do that kind of check in the aggregator rather than here?

Reply:
It doesn't matter whether it's allowed in our installer code, because it's allowed for an aggregated API to be implemented in non-Golang (for simple ones).